packages/jest-cli/src/search_source.js

```diff
@@ -238,4 +239,37 @@ export default class SearchSource {
       return Promise.resolve({tests: []});
     }
   }
+
+  async getTestPaths(
+    globalConfig: GlobalConfig,
+    changedFilesPromise: ?ChangedFilesPromise,
+  ): Promise<SearchResult> {
+    const searchResult = await this._getTestPaths(
+      globalConfig,
+      changedFilesPromise,
+    );
+
+    const filterPath = globalConfig.filter;
+
+    if (filterPath) {
+      const tests = searchResult.tests;
+
+      // $FlowFixMe: dynamic require.
+      const filter = require(filterPath);
+      const filtered = await filter(tests.map(test => test.path));
```

Comment:
Let's change the API to pass the `test` object rather than `test.path`.

Reply:
That was on purpose; the test object carries a lot of additional information that I'm not comfortable exposing to users, like the test time, or access to Haste maps.

Reply:
It would also create all sort of unexpected bugs if people modifies these objects in unexpected ways, like removing a property or something. Sure that's their fault, but the API should rather be strict.